File: 2024-4-14/Palindrome Partitioning II.js
var isPalindrome = function(j,k,string){
    while(j<k){
        if(string[j] !== string[k]){
            return false;
        }
        k--;
        j++
    }
    return true
}
var rec = function(i,len,str,dp){
    if(i===len)return 0;
    if(dp[i] !== -1)return dp[i]
    let min = Number.MAX_VALUE;
    for(let j=i;j<len;j++){
        if(isPalindrome(i,j,str)){
           let cuts =  1+ rec(j+1,len,str,dp)
              min = Math.min(cuts,min)
      }
        
    }
    return dp[i] = min
}
var minCut = function(s) {
    let n = s.length;
    let sol = Array(n).fill(-1);
    return rec(0,n,s,sol) -1
};

File: 2024-4-14/Palindrome Partitioning II.py
#Recursion 
#Time Complexity: O(Exponential)
#Space Complexity: O(n)
class Solution1:
    def minCut(self, s: str) -> int:
        def solve(ind):
            if ind==n:
                return 0
            temp=''
            mini=maxsize
            for j in range(ind,n):
                temp+=s[j]
                if temp==temp[-1::-1]:
                    cuts=1+solve(j+1)
                    mini=min(mini,cuts)
            return mini
        n=len(s)
        return solve(0)-1 